iwyu.cc

```diff
@@ -2553,6 +2553,20 @@ class IwyuBaseAstVisitor : public BaseAstVisitor<Derived> {
     // forward-declarable, even if we're part of a pointer type, or in
     // a template argument, or whatever.
     current_ast_node()->set_in_forward_declare_context(false);
+
+    // When a method is defined outside of a class it has nns as qualifier.
+    // But for method call it doesn't matter if method is defined inside
+    // a class or outside of it.  Skip traversing NestedNameSpecifier
+    // in this case.
+    const ASTNode* ast_node = current_ast_node();
+    if (const CXXMethodDecl* parent = ast_node->GetParentAs<CXXMethodDecl>()) {
+      if ((nns == parent->getQualifier()) &&
+          ast_node->AncestorIsA<CallExpr>(2)) {
```

Comment:
The idea was that when you call a method, it shouldn't matter if it was defined out of line or inside the class. But in other cases method qualifier can be useful and we shouldn't skip visiting it. I tried to remove this check and 9 tests were failing. One of the simplest failing tests is [tests/cxx/no_deque.cc](https://github.com/include-what-you-use/include-what-you-use/blob/master/tests/cxx/no_deque.cc)

```c++
#include <queue>

int main() {
  std::queue<int> q;
  (void)q;
}
```

and without `ast_node->AncestorIsA<CallExpr>(2)` check we recommend to remove `#include <queue>`. Though I didn't check were exactly we go wrong.

But I have some doubts about these lines. I'm not sure about checking parent first and then in  nested block checking node itself and its grandparent. Maybe checking grandparent, then parent, then node itself would be less surprising and easier to understand? There will be more nesting but only 1 extra level and complexity is pretty much the same, just formatted in a different way.